packages/controllers/src/subject-metadata/SubjectMetadataController.ts

```diff
@@ -0,0 +1,261 @@
+  handler: () => void;
+};
+
+export type SubjectMetadataControllerActions =
+  | GetSubjectMetadataState
+  | ClearSubjectMetadataState
+  | TrimSubjectMetadataState;
+
+export type SubjectMetadataStateChange = {
+  type: `${typeof controllerName}:stateChange`;
+  payload: [SubjectMetadataControllerState, Patch[]];
+};
+
+export type SubjectMetadataControllerEvents = SubjectMetadataStateChange;
+
+type AllowedActions = HasPermissions;
+
+export type SubjectMetadataControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  SubjectMetadataControllerActions | AllowedActions,
+  SubjectMetadataControllerEvents,
+  AllowedActions['type'],
+  never
+>;
+
+type SubjectMetadataControllerOptions = {
+  messenger: SubjectMetadataControllerMessenger;
+  subjectCacheLimit: number;
+  state?: Partial<SubjectMetadataControllerState>;
+};
+
+/**
+ * A controller for storing metadata associated with permission subjects. More
+ * or less, a cache.
+ */
+export class SubjectMetadataController extends BaseController<
+  typeof controllerName,
+  SubjectMetadataControllerState,
+  SubjectMetadataControllerMessenger
+> {
+  private _subjectCacheLimit: number;
+
+  get subjectCacheLimit(): number {
+    return this._subjectCacheLimit;
+  }
+
+  private subjectsEncounteredSinceStartup: Set<string>;
+
+  private subjectHasPermissions: GenericPermissionController['hasPermissions'];
+
+  constructor({
+    messenger,
+    subjectCacheLimit,
+    state = {},
+  }: SubjectMetadataControllerOptions) {
+    if (!Number.isInteger(subjectCacheLimit) || subjectCacheLimit < 1) {
+      throw new Error(
+        `subjectCacheLimit must be a positive integer. Received: "${subjectCacheLimit}"`,
+      );
+    }
+
+    const hasPermissions = (origin: string) => {
+      return messenger.call('PermissionController:hasPermissions', origin);
+    };
+
+    if (state.subjectMetadata) {
+      SubjectMetadataController.trimMetadataState(
+        state as SubjectMetadataControllerState,
+        hasPermissions,
+      );
+    }
+
+    super({
+      name: controllerName,
+      metadata: stateMetadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+
+    this.subjectHasPermissions = hasPermissions;
+    this._subjectCacheLimit = subjectCacheLimit;
+    this.subjectsEncounteredSinceStartup = new Set();
+    this.registerMessageHandlers();
+  }
+
+  /**
+   * Constructor helper for registering message handlers for the actions of this
+   * controller.
+   */
+  private registerMessageHandlers(): void {
+    this.messagingSystem.registerActionHandler(
+      `${controllerName}:clearState`,
+      () => this.clearState(),
+    );
+
+    this.messagingSystem.registerActionHandler(
+      `${controllerName}:trimMetadataState`,
+      () => this.trimMetadataState(),
+    );
+  }
+
+  /**
+   * Clears the state of this controller.
+   */
+  clearState(): void {
+    this.update((_draftState) => {
+      return { ...defaultState };
```

Comment:
Letting this get out of sync with `subjectsEncounteredSinceStartup` seems bad. This could lead to a situation where the actual state is getting trimmed to zero because a `clear` left domains in `subjectsEncounteredSinceStartup` equaling the limit.

Reply:
Yeah, good point! fb306e8